audit_trail.go

```diff
@@ -0,0 +1,140 @@
+package tfe
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/google/go-querystring/query"
+	retryablehttp "github.com/hashicorp/go-retryablehttp"
+)
+
+// Compile-time proof of interface implementation
+var _ AuditTrails = (*auditTrails)(nil)
+
+// AuditTrails describes all the audit event related methods that the Terraform
+// Cloud API supports. **Note:** These methods require an organization token for
+// an organization in the Business tier and are only available in Terraform Cloud.
+//
+// TFC API Docs: https://www.terraform.io/cloud-docs/api-docs/audit-trails
+type AuditTrails interface {
+	// Read all the audit events in an organization.
+	List(ctx context.Context, orgToken string, options *AuditTrailListOptions) (*AuditTrailList, error)
+}
+
+// auditTrails implements AuditTrails
+type auditTrails struct {
+	client *Client
+}
+
+// AuditTrailRequest represents the request details of the audit event.
+type AuditTrailRequest struct {
+	ID string `json:"id"`
+}
+
+// AuditTrailAuth represents the details of the actor that invoked the audit event.
+type AuditTrailAuth struct {
+	AccessorID     string  `json:"accessor_id"`
+	Description    string  `json:"description"`
+	Type           string  `json:"type"`
+	ImpersonatorID *string `json:"impersonator_id"`
+	OrganizationID string  `json:"organization_id"`
+}
+
+// AuditTrailResource represents the details of the API resource in the audit event.
+type AuditTrailResource struct {
+	ID     string                  `json:"id"`
+	Type   string                  `json:"type"`
+	Action string                  `json:"action"`
+	Meta   *map[string]interface{} `json:"meta"`
+}
+
+// AuditTrail represents an event in the TFC audit log.
+type AuditTrail struct {
+	ID        string    `json:"id"`
+	Version   string    `json:"version"`
+	Type      string    `json:"type"`
+	Timestamp time.Time `json:"timestamp"`
+
+	Auth     *AuditTrailAuth     `json:"auth"`
+	Request  *AuditTrailRequest  `json:"request"`
+	Resource *AuditTrailResource `json:"resource"`
```

Comment:
What criteria do we use when deciding to use a pointer type for something like this? I tend to prefer "values-by-default" as described in this [blog post](https://medium.com/@meeusdylan/when-to-use-pointers-in-go-44c15fe04eac) and in particular I don't see a compelling reason to use them here (performance, mutability, consistency, etc)

Reply:
This was a great read, thanks. Will change